### DISCUSSION TITLE
Loading custom layout jsons

### DISCUSSION THREAD

**Discussion by Neflardio** [egguy/grott-home-assistant-add-on#61] [Ideas]
Howdy, thanks for the addon! it has been a relief not having to rely on the flaky Growatt servers.

I've noticed Grott has built-in support of loading extra layouts through json files. It would be useful for me if the addon could leverage this feature as I have a situation where the selected standard layout is missing data. Maybe something along the lines of how it loads the grott.ini file and the grott_ha.py file.

**Comment by egguy** [comment]
Hi, you can totaly load any layout.

If you have an editor (vscode / text editor is available as an addon). You can totally change whatever setting you want in `/config/grott` you can even add layout.

**Reply by Neflardio** [reply]
Thanks for letting me know this! Unfortunately I haven't been able to figure out how to add/modify layout via the config. Would you be able to give brief instructions on how to do this?

**Reply by egguy** [reply]
You can't really modify through the config. You need to drop the file in the config directory (/config/grott) and the at the next reload it will load the new overlay.

For this, you can use an existing add-on (VSCode/File Editor) to create the file in the directory (/config/grott/XXXX.json). The format and the name are documented here: https://github.com/johanmeijer/grott/wiki/Grott-advanced-(customize-behaviour).

**Reply by Neflardio** [reply]
Hi sorry for replying to this so late, I have only now just managed to try and load the custom json layout I've created.

Unfortunately my install doesn't seem to pick up my json layout file mthat I've placed in the config directory.

Below is the part of the log where I would expect it to say its loading my custom layout, but as you can see there are no custom layouts found.
```
Feb 13 10:58:01 homeassistant addon_8ce4abae_grott-current[508]: Grott process json layout files                                                                                                                                    
Feb 13 10:58:01 homeassistant addon_8ce4abae_grott-current[508]:                                                                                                                                                                    
Feb 13 10:58:01 homeassistant addon_8ce4abae_grott-current[508]: Grott layout records loaded                                                                                                                                        
Feb 13 10:58:01 homeassistant addon_8ce4abae_grott-current[508]: T02NNNN  :       
```   
Ideally it would pick up the T05NNNNXSPH.json file I've created, as per the log below.

```
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:          - Grott data record length 545
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:          - layout   :  T055004XSPH
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:          - no matching record layout found, try generic
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:          - Record layout used :  T05NNNNXSPH
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:          - Growatt data decrypted V2
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:          - Grott Growatt data decrypted
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:          - Growatt plain data:
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  00b00005021950044e4143353931373438344e544344393238303033000000000000020000007
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  c0006000000040069000100000001006a000300000003000142a009aa07d4089c000000000000
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  000000000000000000000000000000000000000000000000000000000000000000000f54138b0
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  984001400000f5200000000000000000000000000000000000000000000000001030005226d0e
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  1f775b000000ac0002ca96000000af0002e04b000000000000000000000000000000000000000
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  0000000000000000000000000000000000000000000000000000000000005aae1013001180119
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  000002160f4c0f6b0000000000000000000000000000000000000000000000000000000000000
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  00000000000000000010000000000000000000003e80464000600000000000000000020000000
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  000000000015d00000000002120055000007b60000000000000000000007b6000000000000000
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  0000000000000000000001ab9000000000000000000001ab90000010e00060bb3426800000035
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  000184ae0000009500025319000000330001b34a000000720001ae3b00000071000d3e3d00002
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  cf400000000000000000000000000000000000000000000000000000000000003e80000016300
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  000000005514b4fc90001a320051ce59900046000203bb0064163000000000000000460000000
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  000004150415000000cef0cf00cf00cf00cef0cef0cef0cef0cef0cef0cef0cee0cef0cf00cf0
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  0cf000003d02
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:          - Growatt new layout processing
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  - decrypt       :  True
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  - offset        :  6
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:                  - record layout :  T05NNNNXSPH
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]: 
Feb 13 10:58:23 homeassistant addon_8ce4abae_grott-current[508]:          - Grott data record date/time processing started
```

![image](https://github.com/egguy/grott-home-assistant-add-on/assets/13015114/59efb017-efbd-4dfc-a0d8-0f583e651808)

Any help would be much appreciated.

**Comment by Hectorm53** [comment]
Facing the same situation, @Neflardio any achievements so far?

**Reply by Neflardio** [reply]
No achievements unfortunately, don't think the add-on loads stuff from the config/grott folder unfortunately. Tangentially related, but you can have a look at Solar Assistant, which replaces the Growatt wifi completely with a more robust alternative.

**Reply by rhpijnacker** [reply]
Sorry to hear you have not been able to make it work. For me Grott has been running rock-solid for over two years now (every since I moved to my new house). I'm curious to hear why you think Solar Assistant would be a "more robust alternative"?

**Reply by Neflardio** [reply]
Mainly because it replaces the Growatt Shine-wifi. With my installation, the logger loses wireless connection with dongle a few times a year, requiring a reset to reestablish connection.

Beyond that, Solar Assistant allows remote control of the inverter, which can be useful for to tailor usage with ToU energy plans.
